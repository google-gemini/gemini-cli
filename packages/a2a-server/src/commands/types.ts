/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import type { Config } from '@google/gemini-cli-core';

export interface CommandArgument {
  readonly name: string;
  readonly description: string;
  readonly isRequired?: boolean;
}

export interface Command {
  readonly name: string;
  readonly description: string;
  readonly arguments?: CommandArgument[];
  readonly subCommands?: Command[];
  readonly topLevel?: boolean;

  execute(config: Config, args: string[]): Promise<CommandExecutionResponse>;
}

export interface CommandExecutionResponse {
  readonly name: string;
  readonly data: unknown;
}
