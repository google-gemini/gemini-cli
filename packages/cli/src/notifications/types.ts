/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * Defines the structure for individual notification event settings.
 */
export interface NotificationEventSettings {
  enabled: boolean;
  sound: 'system' | 'custom';
  customPath?: string; // Path to custom sound file, only if sound is 'custom'
  timeout?: number; // Optional timeout for idle alerts
}

/**
 * Defines the structure for the overall notification configuration.
 */
export interface NotificationSettings {
  enabled: boolean;
  events: {
    inputRequired: NotificationEventSettings;
    taskComplete: NotificationEventSettings;
    idleAlert: NotificationEventSettings;
  };
}

/**
 * Defines the event types for notifications.
 */
export type NotificationEventType = 'inputRequired' | 'taskComplete' | 'idleAlert';
