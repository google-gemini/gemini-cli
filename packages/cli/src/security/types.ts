/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

// Shared types for MCP server security modules
export type GuardDecision = "allow" | "allow_sanitized" | "escalate" | "block";

export interface GuardResult {
  decision: GuardDecision;
  sanitized?: string;
  reasons: string[];
  tags: string[]; // e.g., ["system-override","tool-coercion","obfuscation"]
}

export interface ContextMeta {
  source: "user" | "rag" | "tool";
  userRole: "user" | "admin" | "service";
  toolAcl: string[];       // allowed tool names
  provenance?: string;     // URL/doc/tool name
  conversationId: string;  // for cross-turn auditing
}

export interface ServerOptions {
  port?: number;
  host?: string;
  policyPath?: string;
  testMode?: boolean; // enable fuzz harness
  enableSecurity?: boolean; // enable security middleware
  enableMetrics?: boolean; // enable real-time metrics
  metricsPort?: number; // port for metrics dashboard
}
