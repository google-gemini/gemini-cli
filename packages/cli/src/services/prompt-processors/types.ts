/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * Defines the interface for a prompt processor, a module that can transform
 * a prompt string before it is sent to the model. Processors are chained
 * together to create a processing pipeline.
 */
export interface IPromptProcessor {
  /**
   * Processes a prompt string, applying a specific transformation.
   *
   * @param prompt The current state of the prompt string.
   * @param args The raw argument string from the user's command invocation.
   * @param fullCommand The full command the user typed (e.g., /git:commit foo).
   * @returns The transformed prompt string as a Promise.
   *
   * @todo Add `context: CommandContext` back when more complex processors
   * that require access to services or UI are needed.
   */
  process(prompt: string, args: string, fullCommand: string): Promise<string>;
}

/**
 * The placeholder string for shorthand argument injection in custom commands.
 */
export const SHORTHAND_ARGS_PLACEHOLDER = '{{args}}';
