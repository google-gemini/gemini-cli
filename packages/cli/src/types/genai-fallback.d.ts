// Minimal local fallbacks for @google/genai types when upstream types aren't available during build.
// These are type-only declarations to keep the consumer compiling. Replace with real upstream types when available.

declare module '@google/genai' {
  // Keep types permissive to avoid leaking runtime coupling in this repo build.
  export type Part = { text?: string; name?: string; type?: string } | string | any;
  export type PartListUnion = Part | Part[] | null;
  export type PartUnion = Part | PartListUnion;
  export type FunctionResponse = { name?: string; arguments?: any } | any;
  export type FunctionCall = { name?: string; args?: any } | any;

  export enum FinishReason {
    FINISH_REASON_UNSPECIFIED = 0,
    STOP = 1,
    MAX_TOKENS = 2,
    SAFETY = 3,
    RECITATION = 4,
    LANGUAGE = 5,
    BLOCKLIST = 6,
    PROHIBITED_CONTENT = 7,
    SPII = 8,
    OTHER = 9,
    MALFORMED_FUNCTION_CALL = 10,
    IMAGE_SAFETY = 11,
    UNEXPECTED_TOOL_CALL = 12,
  }

  export type Content = any;
  export type GenerateContentConfig = any;
}
