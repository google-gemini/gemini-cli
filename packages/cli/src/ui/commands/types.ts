/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { Config, GitService, Logger } from '@google/gemini-cli-core';
import { LoadedSettings } from '../../config/settings.js';
import { HistoryItem, HistoryItemWithoutId, Message } from '../types.js';
import { UseHistoryManagerReturn } from '../hooks/useHistoryManager.js';
import { SessionStatsState } from '../contexts/SessionContext.js';

// Grouped dependencies for clarity and easier mocking
export interface CommandContext {
  // Core services and configuration
  services: {
    config: Config | null;
    settings: LoadedSettings;
    git: GitService | undefined;
    logger: Logger;
  };
  // UI state and history management
  ui: {
    history: HistoryItem[];
    addItem: UseHistoryManagerReturn['addItem'];
    clearItems: UseHistoryManagerReturn['clearItems'];
    loadHistory: UseHistoryManagerReturn['loadHistory'];
    refreshStatic: () => void;
    setQuittingMessages: (messages: HistoryItem[]) => void;
    pendingHistoryItems: HistoryItemWithoutId[];
  };
  // Functions to open dialogs/modals
  dialogs: {
    openTheme: () => void;
    openAuth: () => void;
    openEditor: () => void;
    openPrivacy: () => void;
    setShowHelp: (show: boolean) => void;
  };
  // Specific actions that interact with other hooks/state
  actions: {
    performMemoryRefresh: () => Promise<void>;
    toggleCorgiMode: () => void;
    setPendingCompression: (item: HistoryItemWithoutId | null) => void;
  };
  // Session-specific data
  session: {
    stats: SessionStatsState;
  };
  // Low-level utilities
  utils: {
    onDebugMessage: (message: string) => void;
    addMessage: (message: Message) => void;
  };
}

export interface SlashCommandActionReturn {
  shouldScheduleTool?: boolean;
  toolName?: string;
  toolArgs?: Record<string, unknown>;
  message?: string; // For simple messages or errors
}

// The standardized contract for any command in the system.
export interface SlashCommand {
  name: string;
  altName?: string;
  description?: string;

  // The action to run. Optional for parent commands that only group sub-commands.
  action?: (
    context: CommandContext,
    args: string,
  ) =>
    | void
    | SlashCommandActionReturn
    | Promise<void | SlashCommandActionReturn>;

  // Provides argument completion (e.g., completing a tag for `/chat resume <tag>`).
  completion?: (
    context: CommandContext,
    partialArg: string,
  ) => Promise<string[]>;

  subCommands?: SlashCommand[];
}
