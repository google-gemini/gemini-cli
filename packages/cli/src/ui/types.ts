/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  ToolCallConfirmationDetails,
  ToolResultDisplay,
} from '@gemini-code/server';

// Only defining the state enum needed by the UI
export enum StreamingState {
  Idle = 'idle',
  Responding = 'responding',
  WaitingForConfirmation = 'waiting_for_confirmation',
}

// Copied from server/src/core/turn.ts for CLI usage
export enum GeminiEventType {
  Content = 'content',
  ToolCallRequest = 'tool_call_request',
  // Add other event types if the UI hook needs to handle them
}

export enum ToolCallStatus {
  Pending = 'Pending',
  Canceled = 'Canceled',
  Confirming = 'Confirming',
  Executing = 'Executing',
  Success = 'Success',
  Error = 'Error',
}

export interface ToolCallEvent {
  type: 'tool_call';
  status: ToolCallStatus;
  callId: string;
  name: string;
  args: Record<string, never>;
  resultDisplay: ToolResultDisplay | undefined;
  confirmationDetails: ToolCallConfirmationDetails | undefined;
}

export interface IndividualToolCallDisplay {
  callId: string;
  name: string;
  description: string;
  resultDisplay: ToolResultDisplay | undefined;
  status: ToolCallStatus;
  confirmationDetails: ToolCallConfirmationDetails | undefined;
}

export interface HistoryItemBase {
  text?: string; // Text content for user/gemini/info/error messages
}

// Using Omit<HistoryItem, 'id'> seems to have some issues with typescript's
// type inference e.g. historyItem.type === 'tool_group' isn't auto-inferring that
// 'tools' in historyItem.
export type HistoryItemWithoutId = HistoryItemBase &
  (
    | { type: 'user'; text: string }
    | { type: 'gemini'; text: string }
    | { type: 'gemini_content'; text: string }
    | { type: 'info'; text: string }
    | { type: 'error'; text: string }
    | { type: 'tool_group'; tools: IndividualToolCallDisplay[] }
  );

export type HistoryItem = HistoryItemWithoutId & { id: number };

// Message types used by internal command feedback (subset of HistoryItem types)
export enum MessageType {
  INFO = 'info',
  ERROR = 'error',
  USER = 'user',
  // Add GEMINI if needed by other commands
}

// Simplified message structure for internal feedback
export interface Message {
  type: MessageType;
  content: string; // Renamed from text for clarity in this context
  timestamp: Date; // For consistency, though addItem might use its own timestamping
}
