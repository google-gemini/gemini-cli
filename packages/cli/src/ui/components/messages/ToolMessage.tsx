/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { Box, Text } from 'ink';
import { IndividualToolCallDisplay, ToolCallStatus } from '../../types.js';
import { DiffRenderer } from './DiffRenderer.js';
import { Colors } from '../../colors.js';
import { MarkdownDisplay } from '../../utils/MarkdownDisplay.js';
import { GeminiRespondingSpinner } from '../GeminiRespondingSpinner.js';

const STATIC_HEIGHT = 1;
const RESERVED_LINE_COUNT = 5; // for tool name, status, padding etc.
const STATUS_INDICATOR_WIDTH = 3;

export type TextEmphasis = 'high' | 'medium' | 'low';

export interface ToolMessageProps extends IndividualToolCallDisplay {
  availableTerminalHeight: number;
  emphasis?: TextEmphasis;
}

export const ToolMessage: React.FC<ToolMessageProps> = ({
  name,
  description,
  resultDisplay,
  status,
  availableTerminalHeight,
  emphasis = 'medium',
}) => {
  const contentHeightEstimate =
    availableTerminalHeight - STATIC_HEIGHT - RESERVED_LINE_COUNT;
  const resultIsString =
    typeof resultDisplay === 'string' && resultDisplay.trim().length > 0;
  const lines = React.useMemo(
    () => (resultIsString ? resultDisplay.split('\n') : []),
    [resultIsString, resultDisplay],
  );

  // Truncate the overall string content if it's too long.
  // MarkdownRenderer will handle specific truncation for code blocks within this content.
  // Estimate available height for this specific tool message content area
  // This is a rough estimate; ideally, we'd have a more precise measurement.
  const displayableResult = React.useMemo(
    () =>
      resultIsString
        ? lines.slice(0, contentHeightEstimate).join('\n')
        : resultDisplay,
    [lines, resultIsString, contentHeightEstimate, resultDisplay],
  );
  const hiddenLines = lines.length - contentHeightEstimate;

  return (
    <Box paddingX={1} paddingY={0} flexDirection="column">
      <Box minHeight={1}>
        <ToolStatusIndicator status={status} />
        <ToolInfo
          name={name}
          status={status}
          description={description}
          emphasis={emphasis}
        />
        {emphasis === 'high' && <TrailingIndicator />}
      </Box>
      {displayableResult && (
        <Box paddingLeft={STATUS_INDICATOR_WIDTH} width="100%">
          <Box flexDirection="column">
            {typeof displayableResult === 'string' && (
              <Box flexDirection="column">
                <MarkdownDisplay
                  text={displayableResult}
                  isPending={false}
                  availableTerminalHeight={availableTerminalHeight}
                />
              </Box>
            )}
            {typeof displayableResult !== 'string' && (
              <DiffRenderer
                diffContent={displayableResult.fileDiff}
                filename={displayableResult.fileName}
              />
            )}
            {hiddenLines > 0 && (
              <Box>
                <Text color={Colors.SubtleComment}>
                  ... {hiddenLines} more line{hiddenLines === 1 ? '' : 's'}{' '}
                  hidden ...
                </Text>
              </Box>
            )}
          </Box>
        </Box>
      )}
    </Box>
  );
};

type ToolStatusIndicatorProps = {
  status: ToolCallStatus;
};

const ToolStatusIndicator: React.FC<ToolStatusIndicatorProps> = ({
  status,
}) => (
  <Box minWidth={STATUS_INDICATOR_WIDTH}>
    {status === ToolCallStatus.Pending && (
      <Text color={Colors.AccentGreen}>o</Text>
    )}
    {status === ToolCallStatus.Executing && (
      <GeminiRespondingSpinner
        spinnerType="toggle"
        nonRespondingDisplay={'⊷'}
      />
    )}
    {status === ToolCallStatus.Success && (
      <Text color={Colors.AccentGreen}>✔</Text>
    )}
    {status === ToolCallStatus.Confirming && (
      <Text color={Colors.AccentYellow}>?</Text>
    )}
    {status === ToolCallStatus.Canceled && (
      <Text color={Colors.AccentYellow} bold>
        -
      </Text>
    )}
    {status === ToolCallStatus.Error && (
      <Text color={Colors.AccentRed} bold>
        x
      </Text>
    )}
  </Box>
);

type ToolInfo = {
  name: string;
  description: string;
  status: ToolCallStatus;
  emphasis: TextEmphasis;
};
const ToolInfo: React.FC<ToolInfo> = ({
  name,
  description,
  status,
  emphasis,
}) => {
  const nameColor = React.useMemo<string>(() => {
    switch (emphasis) {
      case 'high':
        return Colors.Foreground;
      case 'medium':
        return Colors.Foreground;
      case 'low':
        return Colors.SubtleComment;
      default: {
        const exhaustiveCheck: never = emphasis;
        return exhaustiveCheck;
      }
    }
  }, [emphasis]);
  return (
    <Box>
      <Text
        wrap="truncate-end"
        strikethrough={status === ToolCallStatus.Canceled}
      >
        <Text color={nameColor} bold>
          {name}
        </Text>{' '}
        <Text color={Colors.SubtleComment}>{description}</Text>
      </Text>
    </Box>
  );
};

const TrailingIndicator: React.FC = () => (
  <Text color={Colors.Foreground}> ←</Text>
);
