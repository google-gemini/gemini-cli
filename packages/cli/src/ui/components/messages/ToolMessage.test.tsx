/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { render } from 'ink-testing-library';
import { ToolMessage, ToolMessageProps } from './ToolMessage.js';
import { StreamingState, ToolCallStatus } from '../../types.js';
import { Text } from 'ink';

// Mock child components or utilities if they are complex or have side effects
vi.mock('ink-spinner', () => ({
  default: () => <Text>MockSpinner</Text>,
}));
vi.mock('./DiffRenderer.js', () => ({
  DiffRenderer: function MockDiffRenderer({
    diffContent,
  }: {
    diffContent: string;
  }) {
    return <Text>MockDiff:{diffContent}</Text>;
  },
}));
vi.mock('../../utils/MarkdownDisplay.js', () => ({
  MarkdownDisplay: function MockMarkdownDisplay({ text }: { text: string }) {
    return <Text>MockMarkdown:{text}</Text>;
  },
}));

describe('<ToolMessage />', () => {
  const baseProps: ToolMessageProps = {
    callId: 'tool-123',
    name: 'test-tool',
    description: 'A tool for testing',
    resultDisplay: 'Test result',
    status: ToolCallStatus.Success,
    availableTerminalHeight: 20,
    confirmationDetails: undefined,
    emphasis: 'medium',
    streamingState: StreamingState.Idle,
  };

  it('renders basic tool information', () => {
    const { lastFrame } = render(<ToolMessage {...baseProps} />);
    const output = lastFrame();
    expect(output).toContain('✔'); // Success indicator
    expect(output).toContain('test-tool');
    expect(output).toContain('A tool for testing');
    expect(output).toContain('MockMarkdown:Test result');
  });

  describe('ToolStatusIndicator rendering', () => {
    it('shows ✔ for Success status', () => {
      const { lastFrame } = render(
        <ToolMessage {...baseProps} status={ToolCallStatus.Success} />,
      );
      expect(lastFrame()).toContain('✔');
    });

    it('shows o for Pending status', () => {
      const { lastFrame } = render(
        <ToolMessage {...baseProps} status={ToolCallStatus.Pending} />,
      );
      expect(lastFrame()).toContain('o');
    });

    it('shows ? for Confirming status', () => {
      const { lastFrame } = render(
        <ToolMessage {...baseProps} status={ToolCallStatus.Confirming} />,
      );
      expect(lastFrame()).toContain('?');
    });

    it('shows - for Canceled status', () => {
      const { lastFrame } = render(
        <ToolMessage {...baseProps} status={ToolCallStatus.Canceled} />,
      );
      expect(lastFrame()).toContain('-');
    });

    it('shows x for Error status', () => {
      const { lastFrame } = render(
        <ToolMessage {...baseProps} status={ToolCallStatus.Error} />,
      );
      expect(lastFrame()).toContain('x');
    });

    it('shows MockSpinner for Executing status when streamingState is Idle', () => {
      const { lastFrame } = render(
        <ToolMessage
          {...baseProps}
          status={ToolCallStatus.Executing}
          streamingState={StreamingState.Idle}
        />,
      );
      expect(lastFrame()).toContain('MockSpinner');
      expect(lastFrame()).not.toContain('✔');
    });

    it('shows MockSpinner for Executing status when streamingState is undefined (default behavior)', () => {
      const { lastFrame } = render(
        <ToolMessage {...baseProps} status={ToolCallStatus.Executing} />,
      );
      expect(lastFrame()).toContain('MockSpinner');
      expect(lastFrame()).not.toContain('✔');
    });

    it('shows ✔ (paused/confirmed look) for Executing status when streamingState is Responding', () => {
      // This is the key change from the commit: if the overall app is still responding
      // (e.g., waiting for other tool confirmations), an already confirmed and executing tool
      // should show a static checkmark to avoid spinner flicker.
      const { lastFrame } = render(
        <ToolMessage
          {...baseProps}
          status={ToolCallStatus.Executing}
          streamingState={StreamingState.Responding} // Simulate app still responding
        />,
      );
      expect(lastFrame()).toContain('✔'); // Should be a checkmark, not spinner
      expect(lastFrame()).not.toContain('MockSpinner');
    });
  });

  it('renders DiffRenderer for diff results', () => {
    const diffResult = {
      fileDiff: '--- a/file.txt\n+++ b/file.txt\n@@ -1 +1 @@\n-old\n+new',
      fileName: 'file.txt',
    };
    const { lastFrame } = render(
      <ToolMessage {...baseProps} resultDisplay={diffResult} />,
    );
    // Check that the output contains the MockDiff content as part of the whole message
    expect(lastFrame()).toMatch(/MockDiff:--- a\/file\.txt/);
  });

  it('renders emphasis correctly', () => {
    const { lastFrame: highEmphasisFrame } = render(
      <ToolMessage {...baseProps} emphasis="high" />,
    );
    // Check for trailing indicator or specific color if applicable (Colors are not easily testable here)
    expect(highEmphasisFrame()).toContain('←'); // Trailing indicator for high emphasis

    const { lastFrame: lowEmphasisFrame } = render(
      <ToolMessage {...baseProps} emphasis="low" />,
    );
    // For low emphasis, the name and description might be dimmed (check for dimColor if possible)
    // This is harder to assert directly in text output without color checks.
    // We can at least ensure it doesn't have the high emphasis indicator.
    expect(lowEmphasisFrame()).not.toContain('←');
  });
});
