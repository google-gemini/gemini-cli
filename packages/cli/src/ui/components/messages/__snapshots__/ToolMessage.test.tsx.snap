// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ToolMessage /> > JSON rendering > renders pretty JSON in ink frame 1`] = `
" âœ“  test-tool A tool for testing

    {
      "a": 1,
      "b": 2
    }"
`;
